sources/packages/api/src/services/application/application.service.ts

```diff
@@ -28,45 +32,210 @@ export class ApplicationService extends RecordDataModelService<Application> {
-    newApplication.data = applicationDto.data;
-    newApplication.studentFiles = studentFiles.map((file) => {
-      const newFileAssociation = new ApplicationStudentFile();
-      newFileAssociation.studentFile = { id: file.id } as StudentFile;
-      return newFileAssociation;
+      `${nextApplicationSequence}`.padStart(sequenceNumberSize, "0")
+    );
+  }
+
+  /**
+   * Saves an Student Application in draft state.
+   * If the application id is provided an update is performed,
+   * otheriwse the draft will be created. The validations are also
+   * apllied accordingly to update/create scenarios.
+   * @param studentId student id.
+   * @param programYearId program year associated with the application draft.
+   * @param applicationData dynamic data received from Form.IO form.
+   * @param associatedFiles associated uploaded files.
+   * @param [applicationId] application id used to execute validations.
+   * @returns Student Application saved draft.
+   */
+  async saveDraftApplication(
+    studentId: number,
+    programYearId: number,
+    applicationData: any,
+    associatedFiles: string[],
+    applicationId?: number,
+  ): Promise<Application> {
+    let draftApplication = await this.getApplicationToSave(
+      studentId,
+      ApplicationStatus.draft,
+    );
+    // If an application id is provided it means that an update is supposed to happen,
+    // so an application draft is expected to be find. If not found, thown an error.
+    if (applicationId && !draftApplication) {
+      throw new CustomNamedError(
+        "Not able to find the draft application associated with the student.",
+        APPLICATION_DRAFT_NOT_FOUND,
+      );
+    }
+    // If an application id is not provided, an insert is supposed to happen
+    // but, if an draft application already exists, it means that it is an
+    // attempt to create a second draft and the student is supposed to
+    // have only one draft.
+    if (!applicationId && draftApplication) {
+      throw new CustomNamedError(
+        "There is already an existing draft application associated with the current student.",
+        ONLY_ONE_DRAFT_ERROR,
+      );
+    }
+    // If an application id is provided, an update is supposed to happen
+    // but, if an draft application already exists under a different program year
+    // the operation should not be allowed.
+    if (draftApplication && draftApplication.programYear.id !== programYearId) {
+      throw new Error(
+        "The application is already saved under a different program year.",
+      );
+    }
+    // If there is no draft application, create one
+    // and initialize the necessary data.
+    if (!draftApplication) {
+      draftApplication = new Application();
+      draftApplication.student = { id: studentId } as Student;
+      draftApplication.programYear = { id: programYearId } as ProgramYear;
+      draftApplication.applicationStatus = ApplicationStatus.draft;
+    }
```

Comment:
TODO: I my current PR, I have introduced `applicationStatusUpdatedOn`, so, will need to add that too  applicationStatusUpdatedOn= getUTCNow();,

Reply:
As we talked about, who merges first will take ensure that this is updated while saving/submitting.